rta/src/main/java/com/gluonhq/richtextarea/viewmodel/RichTextAreaViewModel.java

```diff
@@ -587,24 +588,57 @@ void moveCaret(Direction direction, boolean changeSelection, boolean wordSelecti
     public void resetCharacterIterator() {
         getTextBuffer().resetCharacterIterator();
         updateParagraphList();
-        LOGGER.log(Level.FINE, getTextBuffer().toString());
+        if (LOGGER.isLoggable(Level.FINE)) {
+            LOGGER.log(Level.FINE, getTextBuffer().toString());
+        }
     }
 
     public void walkFragments(BiConsumer<Unit, Decoration> onFragment, int start, int end) {
         getTextBuffer().walkFragments(onFragment, start, end);
     }
 
-    private void updateParagraphList() {
+    void updateParagraphList() {
+        long s0 = System.nanoTime();
         List<Integer> lineFeeds = getTextBuffer().getLineFeeds();
-        AtomicInteger pos = new AtomicInteger();
-        List<Paragraph> newParagraphList = new ArrayList<>();
-        lineFeeds.forEach(lfPos ->
-                newParagraphList.add(getParagraphAt(pos.getAndSet(lfPos), pos.incrementAndGet())));
-        if (pos.get() <= getTextLength()) {
-            lastParagraph = getParagraphAt(pos.get(), getTextLength());
-            newParagraphList.add(lastParagraph);
-        }
-        paragraphList.setAll(newParagraphList);
+        int pos = 0;
+        int counter = 0;
+        int oldSize = paragraphList.size();
+        for (int lfPos : lineFeeds) {
+            int start = pos;
+            int end = lfPos+1;
+            if (counter < oldSize) {
+                Paragraph pg = paragraphList.get(counter);
+                ParagraphDecoration pd = getTextBuffer().getParagraphDecorationAtCaret(start);
+                pg.setStart(start);
```

Comment:
I see the major gain in performance by using a mutable `Paragraph`, instead of doing `paragraphList.set(counter, getParagraphAt(start, end));`